pallets/account/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+
+        /// Claim default evm address for given account id
+        /// Ensure no prior mapping exists for the account
+        #[pallet::call_index(1)]
+        pub fn claim_default_evm_account(origin: OriginFor<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // make sure no prior mapping exists
+            Self::enure_no_mapping(&who, &None)?;
+            // claim default address
+            let _ = Self::do_claim_default_address(who)?;
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Ensure no mappings exists for given pair of account/address
+    fn enure_no_mapping(account_id: &T::AccountId, address: &Option<EvmAddress>) -> DispatchResult {
+        // ensure account_id and address has not been mapped
+        ensure!(
+            !EvmAccounts::<T>::contains_key(&account_id),
+            Error::<T>::AccountIdHasMapped
+        );
+        // This is not required since checking one mapping is sufficent
+        // but this is just for sanity check
+        if let Some(addr) = address {
+            ensure!(
+                !NativeAccounts::<T>::contains_key(addr),
+                Error::<T>::EthAddressHasMapped
+            );
+        }
+        Ok(())
+    }
+
+    /// Add the given pair to create double mappings
+    fn add_mappings(account_id: T::AccountId, address: EvmAddress) {
+        NativeAccounts::<T>::insert(&address, &account_id);
+        EvmAccounts::<T>::insert(&account_id, &address);
+
+        Self::deposit_event(Event::AccountClaimed {
+            account_id,
+            evm_address: address,
+        });
+    }
+
+    /// Claim the given evm address by providing claim signature
+    fn do_claim_address(
+        account_id: T::AccountId,
+        evm_address: EvmAddress,
+        signature: SignatureOf<T>,
+    ) -> DispatchResult {
+        // recover evm address from signature
+        let address = T::ClaimSignature::verify_signature(&account_id, &signature)
+            .ok_or(Error::<T>::BadSignature)?;
+        log::trace!(
+            target: "account::do_claim_address",
+            "evm_address: {:#?}, recovered: {:#?}", evm_address, address
+        );
+        ensure!(evm_address == address, Error::<T>::InvalidSignature);
+
+        // Check if the default account id already exists for this eth address
+        let default_account_id = T::DefaultAddressMapping::into_account_id(evm_address.clone());
+        if frame_system::Pallet::<T>::account_exists(&default_account_id) {
+            // Transfer all the free balance from old account id to the newly
+            // since this `default_account_id` will no longer be connected to evm address
+            // and users cannot access it.
+            T::Currency::transfer(
+                &default_account_id,
+                &account_id,
+                T::Currency::reducible_balance(&default_account_id, Expendable, Polite),
+                Expendable,
+            )?;
+        }
```

Comment:
Didn't we discuss that this should be handled by the UI?

And what about assets in `pallet-assets`?

Reply:
Yes we did, but handling only the native balance from inside pallet is not very high maintenance (one dispatch call) and will in-fact improve the overall UX.
If we don't do this on pallet level then, when user connect the accounts, they first need to sign batch TX to transfer the XC20s and then sign another TX to for moving native balance before finally connecting the account.
The last tx can be skipped with this

The XC20 (pallet-assets) will be handled by UI purely since as we discussed it'll be unnecessarily increase complexity in runtime code.

Reply:
I see - such information should be shared in the PR summary IMO.